Adding single-read functionality to RAW and CLEAN

simonleandergrimm opened the pull request:
This PR adds support for single-read (single-end) sequencing data to the RAW and CLEAN stages of the pipeline while maintaining existing paired-end functionality. This allows the pipeline to process both single-end and paired-end sequencing data using the same workflow infrastructure.



## Key Changes
* Amended generate_samplesheet.sh so it can also take in single-read data.
* Added the run_dev_se.nf workflow, which will be the workflow in which single-read functionality is added up until all steps of the pipeline have single-read functionality. At that point, we can replace run.nf with run_dev_se.nf.
* Added `read_type` parameter ("single_end" or "paired_end") in `run_dev_se.config` to control pipeline behavior
* Split FASTP process into FASTP_SINGLE and FASTP_PAIRED variants
* Split TRUNCATE_CONCAT process into _SINGLE and _PAIRED variants
* Edited RAW, CLEAN, QC, and HV_SCREEN subworkflows to either take in the single_end or paired_end version of processes. 
  * Edited HV_SCREEN only because it would otherwise fail due to not identifying the fastp process correctly.
* Updated `bin/summarize-multiqc-single.R` so it takes in a `read-type` variable, which triggers if/else branches throughout the script to change data processing accordingly. 

## Testing
I added test directories with example data for both single and paired-end cases
  * `test-single-read/` - Contains single-end test data and configuration
  * `test-paired-end/` - Contains paired-end test data and configuration

I validated the pipeline changes in this notebook: https://data.securebio.org/simons-notebook/posts/2024-10-24-mgs-single-read-eval/

willbradshaw commented on .gitignore:
@simonleandergrimm can you sync up with @harmonbhasin re naming here? I think he's going to rename the test directory anyway due to conflict with nf-test.

FWIW I'd prefer something like `test/single/...` and `test/paired/...` to keep the main directory clean.

willbradshaw replied:
Also what are these? 

> analysis_files/*
> mgs-results/

simonleandergrimm replied:
@harmonbhasin What are your thoughts regarding having a test dataset for paired-end and single-end data? Could you rejig your test dataset by e.g., simply keeping the forward reads?

willbradshaw commented on bin/generate_samplesheet.sh:
@harmonbhasin to review changes to this file

simonleandergrimm replied:
@harmonbhasin ping on this.

willbradshaw commented on configs/run.config:
```diff
@@ -5,6 +5,8 @@
     mode = "run"
 
+
+    read_type = "paired_end"
```
This should be unnecessary, we can figure this out based on the format of the sample sheet.

simonleandergrimm replied:
Good idea, will implement.

harmonbhasin replied:
Did this ever get resolved?

willbradshaw commented on test/nextflow.config:
```diff
@@ -5,6 +5,8 @@
 params {
     mode = "run"
 
+    read_type = "paired_end"
```
As above, we shouldn't need this

harmonbhasin replied:
You say implemented, but this pull request still shows you defining read type, am I reviewing the wrong commits?

willbradshaw commented on test-single-read/nextflow.config:
```diff
@@ -0,0 +1,32 @@
+params {
+    mode = "run_dev_se"
+
+    read_type = "single_end"
```
Shouldn't be necessary at all, and if it is it should be a boolean

willbradshaw commented on test-paired-end/nextflow.config:
```diff
@@ -0,0 +1,33 @@
+params {
+    mode = "run_dev_se"
+
+    read_type = "paired_end"
```
See other comments on this

willbradshaw commented on subworkflows/local/raw/main.nf:
```diff
@@ -7,7 +7,12 @@
 include { QC } from "../../../subworkflows/local/qc" addParams(fastqc_cpus: params.fastqc_cpus, fastqc_mem: params.fastqc_mem)
-include { TRUNCATE_CONCAT } from "../../../modules/local/truncateConcat"
+
+if (params.read_type == "single_end") {
```
Should be a boolean flag

willbradshaw commented on bin/summarize-multiqc-single.R:
This will need reworking after pulling in more recent changes from master

willbradshaw commented on configs/run_dev_se.config:
```diff
@@ -0,0 +1,32 @@
+params {
+    mode = "run_dev_se"
+
+    read_type = "single_end"
```
See other comments on this

willbradshaw commented on modules/local/fastp/main.nf:
This is fine, though I wonder if a single process with a boolean `paired` flag would be more elegant.

willbradshaw commented on modules/local/summarizeMultiqcSingle/main.nf:
This will need reworking after pulling in new changes from master

willbradshaw commented on modules/local/truncateConcat/main.nf:
More so than for FASTP, I think this would be better done as a single process with a conditional statement, based either on a boolean `paired` parameter or (better) on the length of `reads`.

You could even just do a for loop iterating over every file in `reads`.

simonleandergrimm replied:
Done

willbradshaw commented on workflows/run_dev_se.nf:
I'm not sure having this serves much purpose when `RUN` is already calling subworkflows that have been modified?

simonleandergrimm replied:
If I interpret your question correctly: run_dev_se.nf exists so the workflow can actually be tested on single-read data. In this branch, anything after RAW/CLEAN will fail, as it's not rejigged to take in single-read data.

harmonbhasin replied:
Ah, this makes sense!

willbradshaw reviewed (changes requested):
Thanks @simonleandergrimm. I left some comments about changes I'd suggest, but the most important one is to pull in the new v2.5.0 code from master and adapt this code to accommodate those changes. Sorry, I know this wouldn't have been necessary with quicker review.

simonleandergrimm commented:
@willbradshaw I've now reworked the code to work with v2.5. The remaining thing we still need to do is settle on the structure of the test datasets. I've assigned the PR to @harmonbhasin and myself to figure this out.

simonleandergrimm commented on bin/generate_samplesheet.sh:
@harmonbhasin can you review this file?

harmonbhasin commented on bin/generate_samplesheet.sh:
This looks good, just wanted to note that this is another reason why #104 .

harmonbhasin commented on configs/run_dev_se.config:
```diff
@@ -0,0 +1,32 @@
+params {
+    mode = "run_dev_se"
+
+    read_type = "single_end"
```
Related to above, shouldn't need to pass in read_type theoretically.

simonleandergrimm replied:
resolved

harmonbhasin commented on configs/run_dev_se.config:
```diff
@@ -0,0 +1,32 @@
+
+    // Directories
+    base_dir = "s3://nao-mgs-simon/test_single_read" // Parent for working and output directories (can be S3)
+    ref_dir = "s3://nao-mgs-wb/index-20240714/output" // Reference/index directory (generated by index workflow)
```
This should be updated

simonleandergrimm replied:
resolved

harmonbhasin commented on test-paired-end/raw/230926Esv_D23-14904-1_1.fastq.gz:
Why do we need to have different test for single/paired? I think we can use the same dataset. Even as we transition into the new test dataset.

simonleandergrimm replied:
Sounds good. Can you point me to the specific changes in https://github.com/naobservatory/mgs-workflow/tree/test-e2e that I can work from to finish this pull request?

simonleandergrimm replied:
Nevermind, I just added the testing regime you suggested at the end of your review.

harmonbhasin commented on test-paired-end/samplesheet.csv:
The way to go will be to have one test directory, which will have a 'paired-samplesheet.csv' and a 'single-samplesheet.csv' or some variant of names.

simonleandergrimm replied:
sgtm

harmonbhasin commented on test-single-read/nextflow.config:
```diff
@@ -0,0 +1,37 @@
+    adapters = "${projectDir}/ref/adapters.fasta" // Path to adapter file for adapter trimming
+
+    // Whether the underlying data is paired-end or single-end
+    single_end = params.sample_sheet.toString().readLines()[0].contains('fastq_2') ? false : true
```
Defining single_end should probably go in the workflow itself. Also why are there two configs that are different 'configs/run_dev_se.config'.

simonleandergrimm replied:
Done

harmonbhasin commented on test-single-read/raw/230926Esv_D23-14904-1.fastq.gz:
Like I said earlier remove.

simonleandergrimm replied:
Done

harmonbhasin commented on test-single-read/samplesheet.csv:
Read earlier comments.

simonleandergrimm replied:
Done

harmonbhasin commented on workflows/run.nf:
I think some funny business is happening where I'm viewing your old commits? If you've already implemented what I commented on, feel free to ignore it.

simonleandergrimm replied:
Yeah I think this is the case. I made some changes to `single-read-raw-profile` that should actually have been propagated to this branch. I will fix.

simonleandergrimm replied:
Resolved

harmonbhasin commented on modules/local/summarizeMultiqcPair/main.nf:
This is personal preference, but I would convert the boolean in the R script.

simonleandergrimm replied:
Done.

harmonbhasin commented on modules/local/summarizeMultiqcPair/resources/usr/bin/summarize-multiqc-pair.R:
```diff
@@ -13,11 +13,14 @@ option_list = list(
               help="Stage descriptor."),
   make_option(c("-S", "--sample"), type="character", default=NULL,
               help="Sample ID."),
+  make_option(c("-r", "--single_end"), type="character", default=FALSE,
```
Shouldn't you make type here boolean if you're doing what you're converting it in the previous script. Noting that I don't know if this is an option for Rparse but I think it should be.

simonleandergrimm replied:
Fixed!

harmonbhasin reviewed (changes requested):
Given that we'll have new tests, the test directory shouldn't matter too much, but I would restructure it for now, or just wait until we push the new tests.

simonleandergrimm commented:
> Given that we'll have new tests, the test directory shouldn't matter too much, but I would restructure it for now, or just wait until we push the new tests.

I don't think it's worth making this change before the new tests. I added an issue to the Project Backlog.

simonleandergrimm commented:
Got back to comments. @harmonbhasin please take another look. FWIW, after positive review by Will, I'd be keen to pull this into dev. Having this feature branch exist alongside `dev` for prolonged periods can create additional work as I need to update the PR when dev changes.

The testing directory for the single-end work needs to be kept separate for now. One can't run `run.nf` on single-end data right now as the processing steps after RAW/CLEAN and PROFILE aren't ready to handle single-end data and will fail. So, as a temporary measure, until full single-end functionality is implemented, I've created separate `run_dev_se.config` and `run_dev_se.nf` files that handle the pipeline from RAW to PROFILE only. Since we have this separate `run_dev_se`, we also need a separate testing directory with a config file that executes the workflow. @harmonbhasin happy to discuss how this fits with your incoming testing regime.

All of this doesn't affect production runs using `run.nf`, as explained in this notebook, which I updated to show results from the newest version of the branch: https://data.securebio.org/simons-notebook/posts/2024-10-24-mgs-single-read-eval/

harmonbhasin commented on bin/generate_samplesheet.sh:
I'm not sure if you're interested in this, but if you want to turn this script into python, I wouldn't be mad lol

simonleandergrimm replied:
👀

harmonbhasin commented on configs/run.config:
I personally believe that this should be moved to the workflow script as it's never going to be changed by the user.

simonleandergrimm replied:
done

harmonbhasin commented on test-dev-se/nextflow.config:
Same comment here about single_end variable.

harmonbhasin reviewed (commented):
This looks good to me, I'm currently testing out some last things with the test workflow, but if things workout well, you should be able to integrate the testing framework tomorrow.

simonleandergrimm commented:
@willbradshaw Pending your review this is ready to go in.

harmonbhasin reviewed (approved):
Looks good to me. I think this is ready for Will.

harmonbhasin commented on .github/workflows/end-to-end.yml:
```diff
@@ -56,6 +57,9 @@ jobs:
       - name: Run run workflow
         run: nf-test test --tag run --verbose
 
+      - name: Run run_dev_se workflow
```
I would make a second github actions test, and call it end-to-end-se.yml, where you only test the run_dev_se. That way end-to-end.yml is standardized amongst different branches.

simonleandergrimm replied:
But don't we want end-to-end.yml test everything? Would this be a temporary measure?

simonleandergrimm replied:
resolved after discussion

harmonbhasin commented on test-data/nextflow.config:
```diff
@@ -7,12 +7,15 @@ params {
     // Directories
     base_dir = "s3://nao-mgs-wb/test-batch" // Parent for working and output directories (can be S3)
-    ref_dir = "s3://nao-mgs-wb/index-20241113/output" // Reference/index directory (generated by index workflow)
+ref_dir = "s3://nao-mgs-wb/index-20241113/output" // Reference/index directory (generated by index workflow)
```
outdated index

simonleandergrimm replied:
waiting for Will to create new index.

simonleandergrimm replied:
Updated

harmonbhasin commented on test-data/nextflow.config:
```diff
@@ -7,12 +7,15 @@ params {
     adapters = "${projectDir}/ref/adapters.fasta" // Path to adapter file for adapter trimming
 
+    // Whether the underlying data is paired-end or single-end
+    single_end = new File(params.sample_sheet).text.readLines()[0].contains('fastq_2') ? false : true
```
you should import the separate config file, right?

simonleandergrimm replied:
yep, fixed!

harmonbhasin commented on workflows/run_dev_se.nf:
```diff
@@ -0,0 +1,96 @@
+    start_time = new Date()
+    start_time_str = start_time.format("YYYY-MM-dd HH:mm:ss z (Z)")
+
+    // Prepare samplesheet
```
Importing the samplesheet is becoming enough lines where I think it may make sense to just put this in a process, perhaps called "LOAD_SAMPLESHEET", and call it in the workflow. @willbradshaw thoughts?

simonleandergrimm replied:
FWIW, I think this would be very good. Happy to create this process.

harmonbhasin replied:
Will said he's good with this, @simonleandergrimm feel free to do this as a subworkflow or a process. The subworkflow will probably be easier as you can write in groovy directly in the subworkflow, but if you can figure out how to do groovy in the process, that works as well.

harmonbhasin commented on modules/local/fastp/main.nf:
```diff
@@ -19,13 +24,23 @@ process FASTP {
         oh=!{sample}_fastp.html
         ad=!{adapters}
-        io="--in1 !{reads[0]} --in2 !{reads[1]} --out1 ${o1} --out2 ${o2} --failed_out ${of} --html ${oh} --json ${oj} --adapter_fasta ${ad}"
+        if [ $(echo "!{reads}" | wc -w) -eq 2 ]; then
```
For some processes, you split the process into two (and call is SINGLE vs PAIRED), whereas in others, you use if statements with in the process (like here for example). What's been your criteria for choosing one over the other? I think we should standardize this if possible.

simonleandergrimm replied:
Roughly depends on how large the changes are. Sometimes, Will had a preference for putting it into the process, sometimes not. Here I actually now switched to having fastp have a single-end and paired-end version.

harmonbhasin replied:
Chatted with Will, going forward, make sure everything is separate, i.e. have a single and a paired process. Thanks for doing that here!

harmonbhasin reviewed (changes requested):
Left some feedback, looking overall pretty good.

willbradshaw commented on configs/read_type.config:
We don't need a whole separate config file just for this parameter.

When I said to make a new config file, I meant for all universal settings. That includes e.g. the stuff at the start and end of `profiles.config`.

Honestly we could put the stuff from `logging.config` and `output.config` in there as well, though that's not essential.

simonleandergrimm replied:
Combining the profiles.config settings with the read type config could cause issues though. Workflows that do not contain a sample sheet but rely on the initial profiles.config settings would probably crash (as samplesheet is undefined)

willbradshaw replied:
If a workflow doesn't define a samplesheet, it will crash anyway when it tries to run LOAD_SAMPLESHEET?

simonleandergrimm replied:
I think we might be talking past each other. What I mean is that a workflow like run_validation would also use the universal_settings.config (as it requires things like logging.config or output.config). If that universal_settings.config contains code that refers to the sample sheet (like `single_end = ....`), run_validation fails.

willbradshaw replied:
Ah, I see. Good point. We can leave this like this for now.

willbradshaw commented on subworkflows/local/loadSampleSheet/main.nf:
```diff
@@ -0,0 +1,45 @@
+| WORKFLOW |
+***********/
+
+workflow LOAD_SAMPLESHET {
```
Typo (here and in places it's called): sampleshet -> samplesheet

simonleandergrimm replied:
quite the feat on me not noticing that typo.

willbradshaw commented on workflows/run_dev_se.nf:
```diff
@@ -0,0 +1,76 @@
+    start_time = new Date()
+    start_time_str = start_time.format("YYYY-MM-dd HH:mm:ss z (Z)")
+
+    // Check if grouping column exists in samplesheet
```
I feel like this should be in LOAD_SAMPLESHEET

simonleandergrimm replied:
Done

willbradshaw commented on workflows/run.nf:
```diff
@@ -40,25 +41,14 @@ workflow RUN {
-        samplesheet_ch = samplesheet.map { sample, read1, read2 -> tuple(sample, [read1, read2]) }
-        group_ch = Channel.empty()
-    }
+    // Load samplesheet
```
I think I would also put the grouping check in here?

simonleandergrimm replied:
Done.

willbradshaw reviewed (changes requested):
Just a few minor changes here, otherwise looks good!

willbradshaw commented:
@simonleandergrimm @harmonbhasin Since we're getting close to merging this would be a good time to update the CHANGELOG.

simonleandergrimm commented:
Aside from Will getting back to my comment and me editing the changelog this is good to go in.

willbradshaw commented on workflows/run_dev_se.nf:
```diff
@@ -51,7 +37,7 @@ workflow RUN_DEV_SE {
     params_str = JsonOutput.prettyPrint(JsonOutput.toJson(params))
     params_ch = Channel.of(params_str).collectFile(name: "run-params.json")
-    time_ch = Channel.of(start_time_str + "\n").collectFile(name: "time.txt")
+    time_ch = Channel.of(params.start_time_str + "\n").collectFile(name: "time.txt")
```
How is this getting into params? Seems naively like it would be simpler to just emit it from LOAD_SAMPLESHEET.

harmonbhasin replied:
@simonleandergrimm what Will is referring to here is this line "params.start_time_str". Previously this was in the main workflow like this:
`start_time = new Date()
    start_time_str = start_time.format("YYYY-MM-dd HH:mm:ss z (Z)")`
. Either move this back to the main workflow (in which case you can use `start_time_str`, or keep it in LOAD_SAMPLESHEET (and make sure to emit it).

harmonbhasin reviewed (changes requested):
@simonleandergrimm outside of the change Will requested, this looks good to me. Make that change and you should be good to go!

simonleandergrimm commented:
@willbradshaw Let me know if this CHANGELOG edit looks good to you. If so I will create the same for the other PR.
```
# v2.5.3
- Added support for single-end read processing:
    - Restructured RAW and CLEAN workflows to handle both single-end and paired-end reads
    - Added new FASTP_SINGLE process alongside existing FASTP_PAIRED
    - Added new TRUNCATE_CONCAT_SINGLE process alongside existing TRUNCATE_CONCAT_PAIRED
    - Added single-end logic to QC and RAW subworkflows
    - Created separate end-to-end test workflow for single-end processing (which will be removed once single-end processing is fully integrated)
- Improved samplesheet handling:
    - Added new LOAD_SAMPLESHEET subworkflow to centralize samplesheet processing
    - Modified samplesheet handling to support both single-end and paired-end data
    - Updated generate_samplesheet.sh to handle single-end data with --single_end flag
- Configuration updates:
    - Added read_type.config to handle single-end vs paired-end settings (set automatically based on samplesheet format)
    - Created run_dev_se.config and run_dev_se.nf for single-end and paired-end development testing (which will be removed once single-end processing is fully integrated)
    - Added single-end samplesheet to test-data
```

willbradshaw commented:
I'm not a huge fan of those CHANGELOG changes because they imply that single-end read processing is more complete than it actually is. I would also keep all the single-end updates under one section:

```
# v2.5.3 (in progress)
- Added new LOAD_SAMPLESHEET subworkflow to centralize samplesheet processing 
- Began development of single-end read processing (still in progress)
    - Restructured RAW, CLEAN, and QC workflows to handle both single-end and paired-end reads
    - Added new FASTP_SINGLE and TRUNCATE_CONCAT_SINGLE processes to handle single-end reads
    - Created separate end-to-end test workflow for single-end processing (which will be removed once single-end processing is fully integrated)
    - Modified samplesheet handling to support both single-end and paired-end data
    - Updated generate_samplesheet.sh to handle single-end data with --single_end flag
    - Added read_type.config to handle single-end vs paired-end settings (set automatically based on samplesheet format)
    - Created run_dev_se.config and run_dev_se.nf for single-end development testing (which will be removed once single-end processing is fully integrated)
    - Added single-end samplesheet to test-data
```

simonleandergrimm commented:
Edited CHANGELOG to incorporate your changes
